test/src/userobjects/ExampleShapeElementUserObject.C

```diff
@@ -0,0 +1,103 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "ExampleShapeElementUserObject.h"
+#include "libmesh/quadrature.h"
+
+template<>
+InputParameters validParams<ExampleShapeElementUserObject>()
+{
+  InputParameters params = validParams<ShapeElementUserObject>();
+  params.addRequiredCoupledVar("u", "first coupled variable");
+  params.addRequiredCoupledVar("v", "second coupled variable");
+  return params;
+}
+
+ExampleShapeElementUserObject::ExampleShapeElementUserObject(const InputParameters & parameters) :
+    ShapeElementUserObject(parameters),
+    _u_value(coupledValue("u")),
+    _u_var(coupled("u")),
+    _v_value(coupledValue("v")),
+    _v_var(coupled("v"))
+{
+}
+
+void
+ExampleShapeElementUserObject::initialize()
+{
+  _integral = 0.0;
+
+  // Jacobian term storage is up to the user. One option is using an std::vector
+  // We resize it to the total number of DOFs in the system and zero it out.
+  // WARNING: this can be large number (smart sparse storage could be a future improvement)
+  // if (_currently_computing_jacobian)
+    _jacobian_storage.assign(_subproblem.es().n_dofs(), 0.0);
```

Comment:
Right now, I am only trying to get the proof of concept, will focus on parallel later. As @dschwen  mentioned for our actual case we only have few non-zero values to assign. However, the test case I am designing will have contribution for all the dofs. I would like to directly store the non-local jacobian values in the sparseMatrix jacobian.